crates/provider/src/shared/provider.rs

```diff
@@ -0,0 +1,63 @@
+use std::error::Error;
+
+use async_trait::async_trait;
+
+use super::types::{NativeRunCommandOptions, PodDef, RunCommandResponse};
+
+#[async_trait]
+#[allow(non_upper_case_globals)]
+pub trait Provider {
+    async fn create_namespace(&mut self) -> Result<(), Box<dyn Error>>;
+    fn get_node_ip(&self) -> Result<String, Box<dyn Error>>;
+    fn get_port_mapping(&mut self, port: u16, pod_name: String) -> Result<u16, Box<dyn Error>>;
+    fn get_node_info(&mut self, pod_name: String) -> Result<(String, u16), Box<dyn Error>>;
+    async fn run_command(
+        &self,
+        args: Vec<String>,
+        opts: NativeRunCommandOptions,
+    ) -> Result<RunCommandResponse, Box<dyn Error>>;
+    async fn run_script(
+        &mut self,
+        identifier: String,
+        script_path: String,
+        args: Vec<String>,
+    ) -> Result<RunCommandResponse, Box<dyn Error>>;
+    fn copy_file_from_pod(
+        &mut self,
+        pod_file_path: String,
+        local_file_path: String,
```

Comment:
Can be relevant to use ```PathBuf``` type from crate ```path``` to encode our requirement in the typesystem